Make ConformanceLookupTable less recursive and don't imply conformances from conditional ones

huonw opened the pull request:
This is preparation for https://bugs.swift.org/browse/SR-6569 along with a proposed fix for rdar://problem/36499373 . This is explicitly not what SE-0143 says, so I'll have to get to amending that before this lands (at least, before the last two commits do).

There was a pile of recursion between `ConformanceLookupTable::updateLookupTable` and the TypeChecker/extension validation, especially for conditional conformances. This takes one approach to breaking that recursion by letting `updateLookupTable` validate much less of the extension until much later. The recursion meant that conformances weren't being constructed when they need to have been. (The commit messages themselves contain a lot more of text/description about exactly what and why things were recurring.)

huonw committed:
[AST] Ignore implied conformance candidates much less aggressively.

The old `existingEntry->getProtocol() == protocol` condition was always
true (the only way for `existingEntry` to get into the array it comes
from is if is exactly for `protocol`), which meant this checking loop
always bailed out, and that there was never more than one implied
candidate ever. This probably isn't what was intended, given there's
more code for handling this case directly after this condition, and
`compareConformances` has extensive code for ranking two implied
conformances for the same nominal/protocol pair.

huonw committed:
[AST] Introduce 'resolveExtensionForConformanceConstruction' for lazier conformance searching.

We need to be able to find which conformances need to be
declared/constructed without forcing extensions to be completely
validated. This is important for both SR-6569 and
rdar://problem/36499373. The former due to the source-level recursion,
and the latter because implied conformances weren't always
constructed (but are needed for good diagnostics).

They weren't always constructed because:

1. ConformanceLookupTable's updateLookupTable on an early stage (before
   implied conformances are found) triggers extension
   validation *before* constructing any conformances, but *after*
   updating the stage state
2. extension validation validates the conditional requirements
3. validating the conditional requirements requires setting up generic
   signatures
4. setting up generic signatures forces the types conformances and so
   ends up in updateLookupTable on the same nominal again, skipping over
   the earlier stages that are complete/in progress
5. we expand the conformances that are implied by all the conformances we
   know about... But we don't know any, because we haven't finished the
   first updateLookupTable.

This breaks the loop at step 2: we instead do the minimal work needed to
know what conformances an extension (might) declare, which is connect
the extension to a type, and then resolve the inherited TypeReprs to
Types.

huonw committed:
[AST] Implied conformances store which conformance implied them.

huonw committed:
[AST] Prefer non-conditional implied conformances.

Given

  protocol P1 {}
  protocol P2: P1 {}
  protocol P3: P1 {}

  struct S<T> {}

  extension S: P2 {}
  extension S: P3 where T: P1 {}

Per SE-0143, we need to make sure that we notice that we can infer S: P1
completely unconditionally from S: P2, and ignore the constrained
one. This is a fairly crude approximation, because doing it accurately
would make this extremely recursive. The approximation used is a
conformance non-conditional if the conformance is from the type, or
there's syntactically no 'where' clause, which is a sufficient condition
for being non-conditional but not necessary (an extension *could* have a
completely redundant 'where' clause).

The recursion problem is checking if something is truly conditional
requires knowing the generic signature of the DeclContext, but knowing
this generic signature requires validating the extension/knowing the
type's conformances, which means hitting this table. (This is made worse
by actual recursive use of conformances, as in SR-6569.)

huonw committed:
[AST] Don't allow conditional conformances to imply others.

Many uses of conditional conformance to protocols with super-protocols
are for wrappers, where the conformances to those super-protocols
usually ends up using weaker bounds, such as:

  struct MyWrapper<Wrapped: Sequence> {}
  extension Wrapped: Sequence {} // Wrapped: Sequence
  extension Wrapped: Collection where Wrapped: Collection {} // *
  extension Wrapped: BidirectionalCollection where Wrapped: BidirectionalCollection {}

If this code was instead written:

  struct MyWrapper<Wrapped: Sequence> {}
  extension Wrapped: Sequence {} // Wrapped: Sequence
  extension Wrapped: BidirectionalCollection where Wrapped: BidirectionalCollection {}

Inferring a Collection conformance would have to mean

  extension Wrapped: Collection where Wrapped: BidirectionalCollection {}

which is unnecessarily strong.

It is a breaking change to change a protocol bound, and so we're
thinking we'd prefer that the compiler didn't magic up that incorrect
conformance. It also only is a small change (and the compiler even
suggests it, with a fixit) to explicitly get the implying behaviour:
declare the conformance explicitly.

  extension Wrapped: Collection, BidirectionalCollection where Wrapped: BidirectionalCollection {}

Fixes rdar://problem/36499373.

huonw commented:
@swift-ci please smoke test

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test source compatibility

DougGregor commented on lib/AST/ConformanceLookupTable.h:
```diff
@@ -372,7 +372,8 @@ class ConformanceLookupTable {
   ProtocolConformance *getConformance(NominalTypeDecl *nominal,
-                                      ConformanceEntry *entry);
+                                      ConformanceEntry *entry,
+                                      LazyResolver *resolver);
```
You can grab the lazy resolver from `nominal->getASTContext().getLazyResolver()`, rather than having to pass it along.

huonw replied:
Oh, I didn't realise that was okay, since the other places here seem to thread it through.

huonw commented on lib/AST/ConformanceLookupTable.cpp:
```diff
@@ -631,6 +631,23 @@ ConformanceLookupTable::Ordering ConformanceLookupTable::compareConformances(
+  auto hasAdditionalRequirements = [&](ConformanceEntry *entry) {
+    if (auto ED = dyn_cast<ExtensionDecl>(entry->getDeclContext()))
+      if (auto TWC = ED->getTrailingWhereClause())
+        return !TWC->getRequirements().empty();
```
Yeah, I guess my comment isn't very clear in that it doesn't explicit state that.

huonw commented on lib/Sema/TypeChecker.cpp:
```diff
@@ -403,6 +403,23 @@ static void bindExtensionDecl(ExtensionDecl *ED, TypeChecker &TC) {
+  TypeResolutionOptions options = TypeResolutionFlags::GenericSignature;
+  options |= TypeResolutionFlags::InheritanceClause;
+  options |= TypeResolutionFlags::AllowUnavailableProtocol;
+  GenericTypeToArchetypeResolver resolver(ext->getInnermostDeclContext());
```
Hm, I guess this should avoid marking things invalid then, given there might be the valid type `SomeClass<T>` there (which is invalid, but the type itself isn't, right?).

I guess using the generic environment of the nominal type isn't correct?

DougGregor replied:
The generic environment of the nominal type would have different archetypes, which would then get recorded in the extension's inheritance list for something like `SomeClass<T>`. You're right that we currently reject something like that further down the line, but it seems like taking the name-binding approach dodges the issue entirely.

huonw commented:
Note to self: CoreStore is failing in the source compat suite. One example is reproduced here for posterity/for when jenkins deletes old runs:

<details>

```
CompileSwift normal x86_64 /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Value.swift
    cd /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore
    /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/build/compat_macos/install/toolchain/usr/bin/swift -frontend -c -filelist /var/folders/_8/79jmzf2142z2xydc_01btlx00000gn/T/sources-37e69a -primary-file /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Value.swift -target x86_64-apple-macosx10.10 -enable-objc-interop -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -I /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Products/Release -F /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Products/Release -application-extension -enable-testing -g -import-underlying-module -module-cache-path /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/ModuleCache.noindex -swift-version 4 -O -disable-swift3-objc-inference -serialize-debugging-options -Xcc "-I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/swift-overrides.hmap" -Xcc -iquote -Xcc "/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/CoreStore-generated-files.hmap" -Xcc "-I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/CoreStore-own-target-headers.hmap" -Xcc "-I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/CoreStore-all-non-framework-target-headers.hmap" -Xcc -ivfsoverlay -Xcc /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/all-product-headers.yaml -Xcc -iquote -Xcc "/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/CoreStore-project-headers.hmap" -Xcc -I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Products/Release/include -Xcc "-I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/DerivedSources/x86_64" -Xcc "-I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/DerivedSources" -Xcc -ivfsoverlay -Xcc "/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/unextended-module-overlay.yaml" -Xcc -working-directory/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore -emit-module-doc-path "/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value~partial.swiftdoc" -serialize-diagnostics-path "/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value.dia" -module-name CoreStore -emit-module-path "/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value~partial.swiftmodule" -emit-dependencies-path "/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value.d" -emit-reference-dependencies-path "/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value.swiftdeps" -o "/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value.o"
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/AttributeProtocol.swift:44:5: warning: 'weak' should not be applied to a property declaration in a protocol and will be disallowed in future versions
    weak var parentObject: CoreStoreObject? { get set }
    ^~~~~
    
Assertion failed: (resolver && "Too late to resolve extensions"), function computeExtensionType, file /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/swift/lib/AST/DeclContext.cpp, line 118.
0  swift                    0x000000010c72b5b8 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift                    0x000000010c72bc46 SignalHandler(int) + 422
2  libsystem_platform.dylib 0x00007fff67c20f5a _sigtramp + 26
3  libsystem_platform.dylib 0x00007fe9ab2fc080 _sigtramp + 1131262272
4  libsystem_c.dylib        0x00007fff67a4b312 abort + 127
5  libsystem_c.dylib        0x00007fff67a13368 basename_r + 0
6  swift                    0x0000000109f36b9d computeExtensionType(swift::ExtensionDecl const*, DeclTypeKind) + 509
7  swift                    0x0000000109f17e8e swift::ConformanceLookupTable::getConformance(swift::NominalTypeDecl*, swift::ConformanceLookupTable::ConformanceEntry*, swift::LazyResolver*) + 190
8  swift                    0x0000000109f18e93 swift::ConformanceLookupTable::getAllConformances(swift::NominalTypeDecl*, swift::LazyResolver*, bool, llvm::SmallVectorImpl<swift::ProtocolConformance*>&) + 163
9  swift                    0x0000000109fc6d75 swift::NominalTypeDecl::getAllConformances(bool) const + 117
10 swift                    0x0000000108dd950c swift::irgen::IRGenModule::addRuntimeResolvableType(swift::NominalTypeDecl*) + 60
11 swift                    0x0000000108de21e0 swift::irgen::IRGenModule::getAddrOfTypeMetadataPattern(swift::NominalTypeDecl*, bool, swift::irgen::ConstantInit, llvm::StringRef) + 288
12 swift                    0x0000000108de1ad0 swift::irgen::IRGenModule::defineTypeMetadata(swift::CanType, bool, bool, bool, clang::CodeGen::ConstantInitFuture, llvm::StringRef) + 112
13 swift                    0x0000000108e4e2d5 swift::irgen::emitClassMetadata(swift::irgen::IRGenModule&, swift::ClassDecl*, swift::irgen::StructLayout const&, swift::irgen::ClassLayout const&) + 4853
14 swift                    0x0000000108dc90b0 swift::irgen::IRGenModule::emitClassDecl(swift::ClassDecl*) + 208
15 swift                    0x0000000108de348c swift::irgen::IRGenModule::emitNestedTypeDecls(swift::IteratorRange<swift::DeclIterator>) + 92
16 swift                    0x0000000108def258 swift::irgen::IRGenModule::emitEnumDecl(swift::EnumDecl*) + 56
17 swift                    0x0000000108dd5c85 swift::irgen::IRGenModule::emitSourceFile(swift::SourceFile&, unsigned int) + 133
18 swift                    0x0000000108eae489 performIRGeneration(swift::IRGenOptions&, swift::ModuleDecl*, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::StringRef, swift::PrimarySpecificPaths const&, llvm::LLVMContext&, swift::SourceFile*, llvm::GlobalVariable**, unsigned int) + 1225
19 swift                    0x0000000108eaeb0e swift::performIRGeneration(swift::IRGenOptions&, swift::SourceFile&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::StringRef, swift::PrimarySpecificPaths const&, llvm::LLVMContext&, unsigned int, llvm::GlobalVariable**) + 94
20 swift                    0x0000000108d3f8d9 performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 15609
21 swift                    0x0000000108d3abce swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3310
22 swift                    0x0000000108cf1623 main + 2051
23 libdyld.dylib            0x00007fff6799f115 start + 1
Stack dump:
0.	Program arguments: /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/build/compat_macos/install/toolchain/usr/bin/swift -frontend -c -filelist /var/folders/_8/79jmzf2142z2xydc_01btlx00000gn/T/sources-37e69a -primary-file /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Value.swift -target x86_64-apple-macosx10.10 -enable-objc-interop -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -I /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Products/Release -F /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Products/Release -application-extension -enable-testing -g -import-underlying-module -module-cache-path /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/ModuleCache.noindex -swift-version 4 -O -disable-swift3-objc-inference -serialize-debugging-options -Xcc -I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/swift-overrides.hmap -Xcc -iquote -Xcc /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/CoreStore-generated-files.hmap -Xcc -I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/CoreStore-own-target-headers.hmap -Xcc -I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/CoreStore-all-non-framework-target-headers.hmap -Xcc -ivfsoverlay -Xcc /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/all-product-headers.yaml -Xcc -iquote -Xcc /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/CoreStore-project-headers.hmap -Xcc -I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Products/Release/include -Xcc -I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/DerivedSources/x86_64 -Xcc -I/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/DerivedSources -Xcc -ivfsoverlay -Xcc /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/unextended-module-overlay.yaml -Xcc -working-directory/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore -emit-module-doc-path /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value~partial.swiftdoc -serialize-diagnostics-path /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value.dia -module-name CoreStore -emit-module-path /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value~partial.swiftmodule -emit-dependencies-path /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value.d -emit-reference-dependencies-path /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value.swiftdeps -o /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/build/Build/Intermediates.noindex/CoreStore.build/Release/CoreStore OSX.build/Objects-normal/x86_64/Value.o 
1.	Contents of /var/folders/_8/79jmzf2142z2xydc_01btlx00000gn/T/sources-37e69a:
---
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSListMonitor.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSCoreStore.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/WhereClauseType.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/ICloudStoreObserver.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSDynamicSchema.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Progress+Convenience.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSEntityDescription+DynamicModel.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSTweak.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/SetupResult.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Progress+ObjectiveC.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSCoreStore+Querying.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStoreObject.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStoreError.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSXcodeDataModelSchema.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/From+Querying.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/QueryableSource.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSSQliteStore.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSGroupBy.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSDataStack+Querying.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/MigrationManager.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSManagedObject+Logging.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/UnsafeDataModelSchema.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSDataStack+Transaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/UserInfo.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSInto.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Value.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSSelect.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSCoreStore+Observing.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSPersistentStore+Setup.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/From.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSCoreStore+Setup.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/DataStack+Transaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/LegacySQLiteStore.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/FetchedResultsControllerDelegate.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/AsynchronousDataTransaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/RelationshipProtocol.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStore+Setup.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStoreStrings.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSFetchedResultsController+Convenience.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSSetupResult.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSFrom.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreDataNativeType.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/ImportableAttributeType.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSListObserver.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStore.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/BaseDataTransaction+Importing.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/XcodeSchemaMappingProvider.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSDataStack+Observing.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStore+Querying.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSError.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSDataStack+Migrating.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/DataStack.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSManagedObjectContext+ObjectiveC.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSManagedObject+Convenience.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/PartialObject.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Where.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSManagedObject+ObjectiveC.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/EntityIdentifier.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSBaseDataTransaction+Querying.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSWhere.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Select.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/SQLiteStore.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSManagedObjectContext+Querying.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSManagedObjectContext+Setup.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NotificationObserver.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStoreFetchRequest+CoreStore.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/DataStack+Querying.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSSectionBy.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/MigrationResult.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSFetchedResultsController+ObjectiveC.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/BaseDataTransaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/SectionBy.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSStorageInterface.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/DynamicSchema+Convenience.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSClauseTypes.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/KeyPath+Querying.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/DataStack+Migration.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSOrderBy.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSEntityDescription+Migration.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/ImportableUniqueObject.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSSynchronousDataTransaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/XcodeDataModelSchema.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSAsynchronousDataTransaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Into.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/StorageInterface.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/FetchableSource.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSSaveResult.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/InMemoryStore.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/GroupBy.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Tweak.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/ImportableObject.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/ObjectMonitor.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/DynamicObject.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/OrderBy.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/MigrationChain.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSManagedObject+Transaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/SchemaHistory.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/SaveResult.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSObjectObserver.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStoreObject+Querying.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/SynchronousDataTransaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/WeakObject.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Functions.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStoreFetchedResultsController.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/QueryChainBuilder.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSCoreStore+Migrating.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStoreSchema.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/ICloudStore.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStore+Logging.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/BaseDataTransaction+Querying.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSInMemoryStore.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/InferredSchemaMappingProvider.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStoreManagedObject.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/FetchChainBuilder.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Relationship.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSManagedObjectContext+CoreStore.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSManagedObjectModel+Migration.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CustomSchemaMappingProvider.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSManagedObject+DynamicModel.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/DispatchQueue+CoreStore.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/DataStack+Observing.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStore+Transaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSUnsafeDataTransaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/ListMonitor.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/ObjectObserver.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/ListObserver.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStoreObject+Convenience.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/UnsafeDataTransaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSCoreStore+Transaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/MigrationType.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Transformable.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSManagedObjectContext+Transaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStore+Observing.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/QueryableAttributeType.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStoreLogger.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/DefaultLogger.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/DynamicKeyPath.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/VersionLock.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSObjectMonitor.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/DynamicSchema.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSUnsafeDataModelSchema.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/SectionMonitorBuilder.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/UnsafeDataTransaction+Observing.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/SchemaMappingProvider.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSMigrationType.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStore+Migration.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSBaseDataTransaction.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSDataStack.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/AttributeProtocol.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStoreBridge.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/NSPersistentStoreCoordinator+Setup.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Entity.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/TypeErasedClauses.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CSMigrationResult.swift
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/CoreStore+CustomDebugStringConvertible.swift
---
2.	While emitting IR for source file /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Value.swift
3.	While emitting class metadata for 'Required' at /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/CoreStore/Sources/Value.swift:76:18
```

</details>

huonw commented:
@swift-ci please smoke test

huonw commented on lib/Sema/TypeChecker.cpp:
```diff
@@ -404,6 +405,36 @@ static void bindExtensionDecl(ExtensionDecl *ED, TypeChecker &TC) {
 void TypeChecker::bindExtension(ExtensionDecl *ext) {
   ::bindExtensionDecl(ext, *this);
 }
+void TypeChecker::resolveExtensionForConformanceConstruction(
```
I've made this more complicated, using `resolveType` and `ResolveStructure` rather than `validateType`. Thoughts?

huonw committed:
[Parse] Put indent-is-4-spaces assumption in one place.

This single location can, theoretically, be made more intelligent about
deducing indent from elsewhere and all the consumers will just work.

huonw commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1447,6 +1447,11 @@ ERROR(objc_generics_cannot_conditionally_conform,none,
       (Type, Type))
+ERROR(conditional_conformances_cannot_imply_conformances,none,
+      "conformance of type %0 to protocol %1 does not imply conformance to "
+      "inherited protocol %2 because it is conditional; explicitly state the "
```
Ah, that's a more succinct phrasing thanks.

huonw commented:
I've changed the non-Swifty fixit to ... three fixits! Similar to `try`/`try!`/`try?`, it suggests three possibly-sensible things, e.g. given `protocol Sub: Base {}` and `extension X: Sub where T: Sub` it offers:

- "relaxed" requirements: `extension X: Base where T: Base`,
- the same requirements: `extension X: Base where T: Sub`,
- different requirements: `extension X: Base where <#requirements#>`.

The relaxed requirements is a heuristic trying to match the common patterns with protocols like `Hashable`/`Equatable`, and the collection hierarchy. It only triggers if *all* the requirements are of the form `<parameter>: Sub`, and they each get rewritten to `<parameter>: Base` (meaning this fixit isn't offered for `extension X: Sub where T == Int` or `extension X: Sub where T: Unrelated`). The other two are offered in all cases.

The full experience of accepting any of them is, for instance:

```swift
extension X: Sub where T: Sub {}
```

&darr;

```swift
extension X: Base where <#requirements#> {
    <#witnesses#>
}

extension X: Sub where T: Sub {}
```

huonw commented:
@swift-ci please smoke test

DougGregor reviewed (approved):
Huon, this looks fantastic! The Fix-Its are great.

huonw commented:
@swift-ci please smoke test

huonw commented:
@swift-ci please smoke test

huonw commented:
@swift-ci smoke test

huonw commented:
@swift-ci please smoke test.

rintaro commented:
Question: Can the fix-it be like this?
```swift
protocol Base {}
protocol Sub: Base {}

struct X<T> {}
```
```swift
extension X: Sub where T: Sub {
}
```
↓
```swift
extension X: Base, Sub where T: Sub {
}
```

huonw commented:
I had that originally, but @DougGregor informed me (https://github.com/apple/swift/pull/15268#discussion_r178177281) that we're trying to move away from that style, and instead prefer having one conformance per extension.

rintaro commented:
Thanks for clarifying. I understand, but the current command line experience is too bad:

```
test.swift:5:1: error: conditional conformance of type 'X<T>' to protocol 'Sub' does not imply conformance to inherited protocol 'Base'
extension X: Sub where T: Sub {
^
test.swift:5:1: note: did you mean to explicitly state the conformance with relaxed bounds?
extension X: Sub where T: Sub {
^
test.swift:5:1: note: did you mean to explicitly state the conformance with the same bounds?
extension X: Sub where T: Sub {
^
test.swift:5:1: note: did you mean to explicitly state the conformance with different bounds?
extension X: Sub where T: Sub {
^
```

Maybe we might want to emit separate messages depending on `DiagnosticsEditorMode`.

huonw commented:
Oh! Do command line diagnostics not include multi-line fixits? That does look bad. I've opened https://bugs.swift.org/browse/SR-7352.
